Source/WebCore/page/NavigateEvent.h

```diff
@@ -0,0 +1,104 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "AbortSignal.h"
+#include "DOMFormData.h"
+#include "Event.h"
+#include "EventInit.h"
+#include "LocalDOMWindowProperty.h"
+#include "NavigationDestination.h"
+#include "NavigationInterceptHandler.h"
+#include "NavigationNavigationType.h"
+
+namespace WebCore {
+
+class NavigateEvent final : public Event {
+    WTF_MAKE_ISO_ALLOCATED(NavigateEvent);
+public:
+    struct Init : EventInit {
+        NavigationNavigationType navigationType { NavigationNavigationType::Push };
+        RefPtr<NavigationDestination> destination;
+        RefPtr<AbortSignal> signal;
+        RefPtr<DOMFormData> formData;
+        String downloadRequest;
+        JSC::JSValue info;
+        bool canIntercept { false };
+        bool userInitiated { false };
+        bool hashChange { false };
+        bool hasUAVisualTransition { false };
+    };
+
+    enum class NavigationFocusReset {
+        AfterTransition,
+        Manual,
+    };
+
+    enum class NavigationScrollBehavior {
+        AfterTransition,
+        Manual,
+    };
+
+    struct NavigationInterceptOptions {
+        RefPtr<NavigationInterceptHandler> handler;
+        NavigationFocusReset focusReset;
+        NavigationScrollBehavior scroll;
+    };
+
+    static Ref<NavigateEvent> create(const AtomString& type, const Init&);
+
+    NavigationNavigationType navigationType() const { return m_navigationType; };
+    bool canIntercept() const { return m_canIntercept; };
+    bool userInitiated() const { return m_userInitiated; };
+    bool hashChange() const { return m_hashChange; };
+    bool hasUAVisualTransition() const { return m_hasUAVisualTransition; };
+    RefPtr<NavigationDestination> destination() { return m_destination; };
+    RefPtr<AbortSignal> signal() { return m_signal; };
+    RefPtr<DOMFormData> formData() { return m_formData; };
+    String downloadRequest() { return m_downloadRequest; };
+    JSC::JSValue info() { return m_info; };
+
+    void intercept(NavigationInterceptOptions&&);
+    void scroll();
+
+private:
+    NavigateEvent(const AtomString& type, const Init&);
+
+    EventInterface eventInterface() const override;
+
+    NavigationNavigationType m_navigationType;
+    RefPtr<NavigationDestination> m_destination;
+    RefPtr<AbortSignal> m_signal;
+    RefPtr<DOMFormData> m_formData;
+    String m_downloadRequest;
+    JSC::JSValue m_info;
+    bool m_canIntercept;
```

Comment:
Let's give these default values in the header to prevent uninitialized memory.
{ false }